utils/src/main/kotlin/ArchiveUtils.kt

```diff
@@ -145,17 +148,18 @@ fun File.unpackDeb(targetDirectory: File) {
             subFolder.safeMkdirs()
 
             val file = tempDir.resolve(name)
-            file.unpack(subFolder)
+            file.unpack(subFolder, filter)
```

Comment:
I presume that the filter acts on the file path in the `deb` package.
The call in line 151 makes it unclear unclear how the filter works, because I guess the filter then now suddenly works on paths relative to the root of the `.xz` library. And there is no means for the caller to make it specific to one `.xz` library.

My preference would be to ditch this commit as it adds complexity and I don't really see the need for it right now. If you strongly think the filter should be done, I won't block it though.

Reply:
It is correct that the filter has no information about the subarchive the entry passed to it is contained. I do not see this as a problem in this specific case, because the paths in the subarchives are disjunct, and even if you do not want to rely on this, this is just a filter; so the worst thing that could happen is that too many files would be extracted.

Our use case is that we want to collect the license information from all the packages contained in a Debian repository. This information is available as a text file contained at a specific path in the package archive. The repository contains ~ 15k packages. So we do not want to extract all of them just to pick up the license files.

Reply:
> It is correct that the filter has no information about the subarchive the entry passed to it is contained. I do not see this as a problem in this specific case, because the paths in the subarchives are disjunct, and even if you do not want to rely on this, this is just a filter; so the worst thing that could happen is that too many files would be extracted.

I don't really see the problem with the overlap which is not happening in reality, but with the (1) broken encapsulation (developer needs to look into the details to understand how the filter actually works and (2) an unclear an misleading function contract, e.g. the developer wonders whereto the paths are relative or falsely assume that the filter paths are relative to the root of the archive. I'd appreciate to see how critical others see that, maybe @oss-review-toolkit/core-devs anyone?

> Our use case is that we want to collect the license information from all the packages contained in a Debian repository. This information is available as a text file contained at a specific path in the package archive. The repository contains ~ 15k packages. So we do not want to extract all of them just to pick up the license files.

How will that use case be integrated into ORT? What component are you planning to add which is responsible for scanning the contained Debian repository?

Reply:
> How will that use case be integrated into ORT? What component are you planning to add which is responsible for scanning the contained Debian repository?

This will be implemented as a Bosch-internal scanner plugin.

Reply:
> I'd appreciate to see how critical others see that, maybe @oss-review-toolkit/core-devs anyone?

I don't see it that critical, mostly because filtering can happen on any property of `ArchiveEntry`, not just the path. For the special case of `unpackDeb()` it would be enough for me to document that `filter()` applies to each nested archive.

Reply:
> This will be implemented as a Bosch-internal scanner plugin.

@oheger-bosch , @sschuberth  if some functionality is being added which is not even planned to be used in ORT, how shall the reviewer review that if he doesn't even know about that fact? (I'm a bit puzzled why this information hasn't been provided with the PR description or commit message)

Reply:
> I'm a bit puzzled why this information hasn't been provided with the PR description or commit message

I was (and still am) convinced that the sole feature to be able to unpack `.deb` archives is valuable to have in ORT upstream. The use in an internal binary scanner is independent of that.

Reply:
> I was (and still am) convinced that the sole feature to be able to unpack `.deb` archives is valuable to have in ORT upstream. The use in an internal binary scanner is independent of that.

Yeah, with just unpacking `.deb` files I didn't see an issue, because it' simple.
With adding the filtering it smelled a bit like unneeded complexity and added redundancy in tests, so I wanted to know the details whether it is really needed. Given the information about use cases in ORT provided, I still cannot see the need for the filtering function. Anyhow, this single case is not a big issue as it''s isolated. But if repeatedly unused code would be added, I'd tend to question that. I'll leave it anyhow up to you guys to decide about this one, I can live with it being merged.

Otherwise LGTM now.

Reply:
> Given the information about use cases in ORT provided, I still cannot see the need for the filtering function.

It's simply a performance optimization: File I/O is slow, esp. when dealing with many files in many archives, and if you're only interested in a small subset of the files, you may want to skip unpacking the other files in the first place rather than ignoring them in a later step.

Reply:
> 
> I don't see it that critical, mostly because filtering can happen on any property of `ArchiveEntry`, not just the path. For the special case of `unpackDeb()` it would be enough for me to document that `filter()` applies to each nested archive.

I have enhanced the documentation of the `unpackDeb()` function accordingly.

Reply:
I believe the added docs mention some unnecessary implementation details. But let's get this merged now, I'll make a follow-up proposal afterwards.